packages/http/src/validator/validators/security/index.ts

```diff
@@ -2,109 +2,54 @@ import { IHttpOperation, HttpSecurityScheme } from '@stoplight/types';
 import { findSecurityHandler } from './handlers';
-import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray';
+import { NonEmptyArray, getSemigroup } from 'fp-ts/lib/NonEmptyArray';
+import { isNonEmpty, array } from 'fp-ts/lib/Array';
 import { IPrismDiagnostic, ValidatorFn } from '@stoplight/prism-core';
 import { IHttpRequest } from '../../../types';
 
-function gatherInvalidResults(
-  error: Either.Left<IPrismDiagnostic>,
-  invalidSecuritySchemes: Array<Array<Either.Either<IPrismDiagnostic, unknown>>>
-) {
-  const invalidSecurity = gatherWWWAuthHeader(invalidSecuritySchemes, ['tags'], error.left);
-  return Option.some(invalidSecurity);
-}
-
-function gatherValidationResults(
-  securitySchemes: HttpSecurityScheme[][],
-  input: Pick<IHttpRequest, 'headers' | 'url'>
-) {
-  const authResults = getAuthResults(securitySchemes, input);
-
-  const validSecurityScheme = authResults.find(authRes => authRes.every(Either.isRight));
-  const invalidSecuritySchemes = authResults.filter(authRes => authRes.some(Either.isLeft));
-
-  const firstLeft = invalidSecuritySchemes[0] && invalidSecuritySchemes[0].find(Either.isLeft);
-
-  if (!validSecurityScheme && firstLeft) {
-    return gatherInvalidResults(firstLeft, invalidSecuritySchemes);
-  } else {
-    return Option.none;
-  }
-}
-
-function gatherWWWAuthHeader(
-  authResults: Array<Array<Either.Either<IPrismDiagnostic, unknown>>>,
-  pathToHeader: string[],
-  firstAuthErr: IPrismDiagnostic
-) {
-  const flattenedAuthResults = flatten(authResults);
-
-  if (flattenedAuthResults.length === 1) {
-    return firstAuthErr;
-  } else {
-    const wwwAuthenticateHeaders = flattenedAuthResults.map(authResult =>
-      pipe(
-        authResult,
-        Either.fold(result => result.tags || [], noop)
-      )
-    );
+type HeadersAndUrl = Pick<IHttpRequest, 'headers' | 'url'>;
 
-    const firstAuthErrWithAuthHeader = set(pathToHeader, flatten(wwwAuthenticateHeaders), firstAuthErr);
+const EitherValidation = Either.getValidation(getSemigroup<IPrismDiagnostic>());
+const eitherSequence = array.sequence(EitherValidation);
 
-    return wwwAuthenticateHeaders.every(identity) ? firstAuthErrWithAuthHeader : firstAuthErr;
-  }
+function getValidationResults(securitySchemes: HttpSecurityScheme[][], input: HeadersAndUrl) {
+  const [first, ...others] = getAuthResultsAND(securitySchemes, input);
+  return others.reduce((prev, current) => EitherValidation.alt(prev, () => current), first);
 }
 
-function getAuthResult(
-  firstAuthErrAsLeft: Either.Left<IPrismDiagnostic>,
-  authResult: Array<Either.Either<IPrismDiagnostic, unknown>>
-) {
-  const firstAuthErr = pipe(
-    firstAuthErrAsLeft,
-    Either.fold<IPrismDiagnostic, IPrismDiagnostic, IPrismDiagnostic>(identity, identity)
-  );
-
-  const invalidResultWithAuthHeader = gatherWWWAuthHeader([authResult], ['tags'], firstAuthErr);
-
-  return [Either.left(invalidResultWithAuthHeader)];
+function getWWWAuthHeader(authResults: NonEmptyArray<IPrismDiagnostic>) {
```

Comment:
I have trouble understanding correlation between the name of the function and what it actually does.

It takes a list of diagnostics, merges the tags lists from them and writes that to the first diagnostic's tags field.

How is the above connected with `WWWAuthHeader`?

Reply:
That's a very good point — this indeed produces *what* gets put in the WWWAuthHeader returned to the client.